spark_pipeline_framework/utilities/file_downloader.py

```diff
@@ -0,0 +1,137 @@
+import os
+import shutil
+import tempfile
+from typing import Any, Optional
+from urllib import request as url_request
+from urllib import parse as url_parse
+from zipfile import ZipFile, is_zipfile
+
+
+class ThrowOnErrorOpener(url_request.FancyURLopener):
+    def http_error_default(
+        self, url: str, fp: Any, errcode: int, errmsg: str, headers: Any
+    ) -> Any:
+        raise Exception(f"{errcode}: {errmsg}")
+
+
+class FileDownloader:
+    """
+    Module to download files from urls
+    Inspired from python wget module: https://github.com/steveeJ/python-wget
+    """
+
+    def __init__(
+        self, url: str, download_path: Optional[str] = "",
+    ):
+        self.url = url
+        self.download_to_path = download_path or ""
+
+    @staticmethod
+    def check_if_path_exists(path: str) -> bool:
+        if not os.path.exists(path):
+            raise FileNotFoundError(f"File does not exists at this path: {path}")
+        return True
+
+    @staticmethod
+    def get_filename_from_url(url: str) -> str:
+        filename = os.path.basename(url_parse.urlparse(url).path)
+        filename = filename.strip(" \n\t.")
+        if len(filename) == 0:
+            return None  # type: ignore
+        return filename
+
+    # @staticmethod
+    # def get_filename_from_headers(headers: Any) -> str:
+    #     """Detect filename from Content-Disposition headers if present.
+    #     :param headers: Dict/List/str:
+    #
+    #     """
+    #     if isinstance(headers, str):
+    #         headers = headers.splitlines()
+    #     if isinstance(headers, list):
+    #         headers = dict([x.split(':', 1) for x in headers])
+    #     cdisp = headers.get("Content-Disposition")
+    #     if not cdisp:
+    #         return None
+    #     cdtype = cdisp.split(';')
+    #     if len(cdtype) == 1:
+    #         return None
+    #     if cdtype[0].strip().lower() not in ('inline', 'attachment'):
+    #         return None
+    #     # several filename params is illegal, but just in case
+    #     fnames = [x for x in cdtype[1:] if x.strip().startswith('filename=')]
+    #     if len(fnames) > 1:
+    #         return None
+    #     name = fnames[0].split('=')[1].strip(' \t"')
+    #     name = os.path.basename(name)
+    #     if not name:
+    #         return None
+    #     return name
+
+    @staticmethod
+    def rename_filename_if_exists(filename: str) -> str:
```

Comment:
I suggest GUID based method, this looks unnecessarily complicated

Reply:
GUID is a great way of downloading files, but I would be needing some extra effort to track the name mapping of multiple files as I need them properly named before uploading them to the s3 datalake.